### ISSUE TITLE
Feature request: import show hints

### ISSUE THREAD

**Issue by Wdestroier** [Dart-Code/Dart-Code#5240]
**Is your feature request related to a problem? Please describe.**

In Java we can see all imported symbols, but the import string is large. For example:
```java
import java.lang.String;
import java.lang.StringBuilder;
```
In Dart we have shorter imports:
```dart
import 'dart:core';
```
We can't easily see which symbols are imported from `'dart:core'`.

**Describe the solution you'd like**

I would like to see the IDE always showing a greyed out hint like "`show num, int;`". Example of a greyed out commit name:
![image](https://github.com/user-attachments/assets/e134bb73-4488-4f5c-92fc-28c913a5e6ce)

The hint would show all symbols from the imported library which are used in the current library. The hint would not show all exported symbols from the imported library.

**Describe alternatives you've considered**
I tried to make the IDE always add imports in the show clause, but it was a bad experience, #5238 is one of the problems.

I tried to delete the import line and check every error in the file, but many symbols exist in many places on the code.

**Comment by DanTup** [comment]
This may be possible, but I think the lists could be very long for some files so I don't know how the experience would be.

If we were to try this out, it would be better if we could do it generally using InlayHints in the LSP server rather than something custom (so it can work for any LSP client), but currently VS Code doesn't support inlay hints that aren't considered Types or Parameters - see https://github.com/microsoft/vscode/issues/151920.

**Comment by DanTup** [comment]
The hover code is _mostly_ shared with IntelliJ, yep.

I'm not sure if showing all symbols from an import in the hover (or an inlay hint as I suggested above) would be a great experience though.

There are a number of open requests related to hovers and it might make sense to review hovers more generally before just adding more things to it for some specific kinds of hovers.

**Comment by Wdestroier** [comment]
The bug mentioned in the issue description (which is already fixed) is a better solution than these hints imo. I still import `material.dart` without a show clause, because this file has too many exports. The hint can be useful for these kind of files with too many exports imo.

> I'm not sure if showing all symbols from an import in the hover (or an inlay hint as I suggested above) would be a great experience though.

I dislike the idea of displaying the show clause on hover, because hovering each import trying to find a specific class name would be too much work.

**Comment by FMorschel** [comment]
> hovering each import trying to find a specific class name

Are you aware of:

- https://github.com/Dart-Code/Dart-Code/issues/5326

Take a look here https://dartcode.org/releases/#v31020-2024-12-02 and scroll down just a bit to see it working.

**Comment by Wdestroier** [comment]
> Are you aware of: https://github.com/Dart-Code/Dart-Code/issues/5326

Cool! I just tried in Cursor and it works.

A show hint could be nice to see all imports coming from dart/async or flutter/material. An example import for a single `SfDataGrid` widget from another package:
![Image](https://github.com/user-attachments/assets/e455281b-68c0-47b2-860a-da747461d611)
I didn't need to import too many symbols from this package, but Flutter has too many symbols. About 1 import for each line of code, excluding closing braces.

> I'm not sure if showing all symbols from an import in the hover (or an inlay hint as I suggested above) would be a great experience though.

Maybe the show-clause inlay hints could be an option people can enable. Separated from the "type info" inlay hints, which get in the way when trying to write code. I'm unsure if VSCode provides the APIs to implement this.

**Comment by Wdestroier** [comment]
This issue is very interesting! I gave it a thumbs up.

**Comment by DanTup** [comment]
> The bug mentioned in the issue description (which is already fixed) is a better solution than these hints imo. 

Do you mean this issue is no longer a problem now that's fixed?

My concern with doing this is that many libraries export enormous numbers of symbols, so those lists would be very long and need updating every time the file changes. 

> Take a look at:
> 
> https://github.com/dart-lang/sdk/issues/60326

I'll note that that issue is about Inlay Hints, and we can't currently use Inlay Hints for anything except types and parameters because VS Code has a fixed set of `kind`s (see https://github.com/microsoft/vscode/issues/151920).

**Comment by FMorschel** [comment]
> I'll note that that issue is about Inlay Hints, and we can't currently use Inlay Hints for anything except types and parameters because VS Code has a fixed set of kinds (see https://github.com/microsoft/vscode/issues/151920).

Yes, sorry if I wasn't explicit about this. Only mentioned because the current suggestion was to use this feature when this VS Code issue is closed (you mentioned it above) and for this to be optional this SDK issue should be closed too.

> My concern with doing this is that many libraries export enormous numbers of symbols, so those lists would be very long and need updating every time the file changes.

I'm aware that exporting many symbols may be a problem in processing but from what I gathered this was supposed to display _used_ symbols from what I could gather.

---

I'll open an issue on the SDK about open/close for combinators later today.

**Comment by FMorschel** [comment]
> I'll open an issue on the SDK about open/close for combinators later today.

Here it is:

- https://github.com/dart-lang/sdk/issues/60370

**Comment by Wdestroier** [comment]
> I'm aware that exporting many symbols may be a problem in processing but from what I gathered this was supposed to display used symbols from what I could gather.

Yes, only referenced symbols would be shown by the hint.
